File: apps/flutter/lib/services/desktop_menu_service.dart
import 'dart:io';
import 'package:flutter/material.dart';
import 'package:flutter/services.dart';

/// Desktop menu bar service for macOS, Windows, and Linux
class DesktopMenuService {
  /// Check if we're on a desktop platform
  static bool get isDesktop {
    return Platform.isMacOS || Platform.isWindows || Platform.isLinux;
  }

  /// Helper to create platform-appropriate shortcut
  static SingleActivator _shortcut(LogicalKeyboardKey key, {bool shift = false}) {
    if (Platform.isMacOS) {
      return SingleActivator(key, meta: true, shift: shift);
    } else {
      return SingleActivator(key, control: true, shift: shift);
    }
  }

  /// Build the platform menu bar
  static List<PlatformMenuItem> buildMenuBar({
    required VoidCallback onSearch,
    required VoidCallback onSettings,
    required VoidCallback onFilters,
    required VoidCallback onRefresh,
    required VoidCallback onExport,
    required VoidCallback onPrint,
    required VoidCallback onToggleTheme,
    required VoidCallback onGoHome,
    required VoidCallback onGoSaved,
    required VoidCallback onQuit,
    required VoidCallback onAbout,
  }) {
    return [
      // App Menu (macOS only - contains About and Quit)
      if (Platform.isMacOS)
        PlatformMenu(
          label: 'Bay Navigator',
          menus: [
            PlatformMenuItem(
              label: 'About Bay Navigator',
              onSelected: onAbout,
            ),
            PlatformMenuItem(
              label: 'Quit Bay Navigator',
              shortcut: _shortcut(LogicalKeyboardKey.keyQ),
              onSelected: onQuit,
            ),
          ],
        ),

      // File Menu
      PlatformMenu(
        label: 'File',
        menus: [
          PlatformMenuItem(
            label: 'Export Saved Programs...',
            shortcut: _shortcut(LogicalKeyboardKey.keyE, shift: true),
            onSelected: onExport,
          ),
          PlatformMenuItem(
            label: 'Print Saved Programs...',
            shortcut: _shortcut(LogicalKeyboardKey.keyP),
            onSelected: onPrint,
          ),
          if (!Platform.isMacOS)
            PlatformMenuItem(
              label: 'Exit',
              shortcut: _shortcut(LogicalKeyboardKey.keyQ),
              onSelected: onQuit,
            ),
        ],
      ),

      // Edit Menu
      PlatformMenu(
        label: 'Edit',
        menus: [
          PlatformMenuItem(
            label: 'Search Programs',
            shortcut: _shortcut(LogicalKeyboardKey.keyF),
            onSelected: onSearch,
          ),
          PlatformMenuItem(
            label: 'Filter Programs',
            shortcut: _shortcut(LogicalKeyboardKey.keyL),
            onSelected: onFilters,
          ),
        ],
      ),

      // View Menu
      PlatformMenu(
        label: 'View',
        menus: [
          PlatformMenuItem(
            label: 'Refresh',
            shortcut: _shortcut(LogicalKeyboardKey.keyR),
            onSelected: onRefresh,
          ),
          PlatformMenuItem(
            label: 'Toggle Dark Mode',
            shortcut: _shortcut(LogicalKeyboardKey.keyD, shift: true),
            onSelected: onToggleTheme,
          ),
        ],
      ),

      // Go Menu
      PlatformMenu(
        label: 'Go',
        menus: [
          PlatformMenuItem(
            label: 'Home',
            shortcut: _shortcut(LogicalKeyboardKey.digit1),
            onSelected: onGoHome,
          ),
          PlatformMenuItem(
            label: 'Saved Programs',
            shortcut: _shortcut(LogicalKeyboardKey.digit2),
            onSelected: onGoSaved,
          ),
          PlatformMenuItem(
            label: 'Settings',
            shortcut: _shortcut(LogicalKeyboardKey.comma),
            onSelected: onSettings,
          ),
        ],
      ),

      // Help Menu (About is here for Windows/Linux, in App menu for macOS)
      PlatformMenu(
        label: 'Help',
        menus: [
          if (!Platform.isMacOS)
            PlatformMenuItem(
              label: 'About Bay Navigator',
              onSelected: onAbout,
            ),
        ],
      ),
    ];
  }
}

File: apps/apple/BayNavigatorCore/Sources/BayNavigatorCore/Services/DesktopMenuService.swift
import Foundation
import SwiftUI

/// Desktop menu bar service for macOS
/// Provides the standard macOS menu bar with keyboard shortcuts
#if os(macOS)

// MARK: - Menu Commands

/// Standard app commands for Bay Navigator
public struct BayNavigatorCommands: Commands {
    // Actions
    let onSearch: () -> Void
    let onSettings: () -> Void
    let onFilters: () -> Void
    let onRefresh: () -> Void
    let onExport: () -> Void
    let onPrint: () -> Void
    let onToggleTheme: () -> Void
    let onGoToTab: (Int) -> Void
    let onShowKeyboardShortcuts: () -> Void

    public init(
        onSearch: @escaping () -> Void = {},
        onSettings: @escaping () -> Void = {},
        onFilters: @escaping () -> Void = {},
        onRefresh: @escaping () -> Void = {},
        onExport: @escaping () -> Void = {},
        onPrint: @escaping () -> Void = {},
        onToggleTheme: @escaping () -> Void = {},
        onGoToTab: @escaping (Int) -> Void = { _ in },
        onShowKeyboardShortcuts: @escaping () -> Void = {}
    ) {
        self.onSearch = onSearch
        self.onSettings = onSettings
        self.onFilters = onFilters
        self.onRefresh = onRefresh
        self.onExport = onExport
        self.onPrint = onPrint
        self.onToggleTheme = onToggleTheme
        self.onGoToTab = onGoToTab
        self.onShowKeyboardShortcuts = onShowKeyboardShortcuts
    }

    public var body: some Commands {
        // Replace the default New/Open with our custom File menu items
        CommandGroup(replacing: .newItem) {
            // Empty - we don't need New Document
        }

        // File Menu - Export and Print
        CommandGroup(after: .newItem) {
            Button("Export Saved Programs...") {
                onExport()
            }
            .keyboardShortcut("e", modifiers: [.command, .shift])

            Button("Print Saved Programs...") {
                onPrint()
            }
            .keyboardShortcut("p", modifiers: .command)
        }

        // Edit Menu - Search and Filters
        CommandGroup(after: .pasteboard) {
            Divider()

            Button("Search Programs") {
                onSearch()
            }
            .keyboardShortcut("f", modifiers: .command)

            Button("Filter Programs") {
                onFilters()
            }
            .keyboardShortcut("l", modifiers: .command)
        }

        // View Menu
        CommandGroup(after: .toolbar) {
            Button("Refresh") {
                onRefresh()
            }
            .keyboardShortcut("r", modifiers: .command)

            Divider()

            Button("Toggle Dark Mode") {
                onToggleTheme()
            }
            .keyboardShortcut("d", modifiers: [.command, .shift])
        }

        // Custom Go Menu
        CommandMenu("Go") {
            Button("Home") {
                onGoToTab(0)
            }
            .keyboardShortcut("1", modifiers: .command)

            Button("Saved Programs") {
                onGoToTab(1)
            }
            .keyboardShortcut("2", modifiers: .command)

            Button("More") {
                onGoToTab(2)
            }
            .keyboardShortcut("3", modifiers: .command)

            Divider()

            Button("Settings") {
                onSettings()
            }
            .keyboardShortcut(",", modifiers: .command)
        }

        // Help Menu
        CommandGroup(replacing: .help) {
            Button("Keyboard Shortcuts") {
                onShowKeyboardShortcuts()
            }
            .keyboardShortcut("/", modifiers: .command)

            Divider()

            Link("Bay Navigator Help",
                 destination: URL(string: "https://baynavigator.org/help")!)

            Link("Report an Issue",
                 destination: URL(string: "https://github.com/baytides/baynavigator/issues")!)

            Divider()

            Link("Privacy Policy",
                 destination: URL(string: "https://baynavigator.org/privacy")!)

            Link("Terms of Service",
                 destination: URL(string: "https://baynavigator.org/terms")!)
        }
    }
}

// MARK: - Menu Bar State

/// Observable state for menu bar actions
@MainActor
public final class MenuBarState: ObservableObject {
    public static let shared = MenuBarState()

    // Published actions that views can subscribe to
    @Published public var searchTriggered = false
    @Published public var settingsTriggered = false
    @Published public var filtersTriggered = false
    @Published public var refreshTriggered = false
    @Published public var exportTriggered = false
    @Published public var printTriggered = false
    @Published public var themeToggleTriggered = false
    @Published public var keyboardShortcutsTriggered = false
    @Published public var navigateToTab: Int?

    private init() {}

    // MARK: - Action Triggers

    public func triggerSearch() {
        searchTriggered = true
        // Reset after a brief delay
        Task { @MainActor in
            try? await Task.sleep(nanoseconds: 100_000_000)
            searchTriggered = false
        }
    }

    public func triggerSettings() {
        settingsTriggered = true
        Task { @MainActor in
            try? await Task.sleep(nanoseconds: 100_000_000)
            settingsTriggered = false
        }
    }

    public func triggerFilters() {
        filtersTriggered = true
        Task { @MainActor in
            try? await Task.sleep(nanoseconds: 100_000_000)
            filtersTriggered = false
        }
    }

    public func triggerRefresh() {
        refreshTriggered = true
        Task { @MainActor in
            try? await Task.sleep(nanoseconds: 100_000_000)
            refreshTriggered = false
        }
    }

    public func triggerExport() {
        exportTriggered = true
        Task { @MainActor in
            try? await Task.sleep(nanoseconds: 100_000_000)
            exportTriggered = false
        }
    }

    public func triggerPrint() {
        printTriggered = true
        Task { @MainActor in
            try? await Task.sleep(nanoseconds: 100_000_000)
            printTriggered = false
        }
    }

    public func triggerThemeToggle() {
        themeToggleTriggered = true
        Task { @MainActor in
            try? await Task.sleep(nanoseconds: 100_000_000)
            themeToggleTriggered = false
        }
    }

    public func triggerKeyboardShortcuts() {
        keyboardShortcutsTriggered = true
        Task { @MainActor in
            try? await Task.sleep(nanoseconds: 100_000_000)
            keyboardShortcutsTriggered = false
        }
    }

    public func navigateTo(tab: Int) {
        navigateToTab = tab
        Task { @MainActor in
            try? await Task.sleep(nanoseconds: 100_000_000)
            navigateToTab = nil
        }
    }
}

// MARK: - Convenience Commands Factory

/// Factory for creating pre-configured menu commands
public struct DesktopMenuService {
    /// Create commands that use the shared MenuBarState
    public static func createCommands() -> BayNavigatorCommands {
        let state = MenuBarState.shared

        return BayNavigatorCommands(
            onSearch: { state.triggerSearch() },
            onSettings: { state.triggerSettings() },
            onFilters: { state.triggerFilters() },
            onRefresh: { state.triggerRefresh() },
            onExport: { state.triggerExport() },
            onPrint: { state.triggerPrint() },
            onToggleTheme: { state.triggerThemeToggle() },
            onGoToTab: { tab in state.navigateTo(tab: tab) },
            onShowKeyboardShortcuts: { state.triggerKeyboardShortcuts() }
        )
    }
}

// MARK: - View Modifier for Menu Actions

/// View modifier that responds to menu bar actions
public struct MenuBarActionHandler: ViewModifier {
    @ObservedObject private var menuState = MenuBarState.shared

    let onSearch: () -> Void
    let onSettings: () -> Void
    let onFilters: () -> Void
    let onRefresh: () -> Void
    let onExport: () -> Void
    let onPrint: () -> Void
    let onToggleTheme: () -> Void
    let onGoToTab: (Int) -> Void
    let onShowKeyboardShortcuts: () -> Void

    public init(
        onSearch: @escaping () -> Void = {},
        onSettings: @escaping () -> Void = {},
        onFilters: @escaping () -> Void = {},
        onRefresh: @escaping () -> Void = {},
        onExport: @escaping () -> Void = {},
        onPrint: @escaping () -> Void = {},
        onToggleTheme: @escaping () -> Void = {},
        onGoToTab: @escaping (Int) -> Void = { _ in },
        onShowKeyboardShortcuts: @escaping () -> Void = {}
    ) {
        self.onSearch = onSearch
        self.onSettings = onSettings
        self.onFilters = onFilters
        self.onRefresh = onRefresh
        self.onExport = onExport
        self.onPrint = onPrint
        self.onToggleTheme = onToggleTheme
        self.onGoToTab = onGoToTab
        self.onShowKeyboardShortcuts = onShowKeyboardShortcuts
    }

    public func body(content: Content) -> some View {
        content
            .onChange(of: menuState.searchTriggered) { _, triggered in
                if triggered { onSearch() }
            }
            .onChange(of: menuState.settingsTriggered) { _, triggered in
                if triggered { onSettings() }
            }
            .onChange(of: menuState.filtersTriggered) { _, triggered in
                if triggered { onFilters() }
            }
            .onChange(of: menuState.refreshTriggered) { _, triggered in
                if triggered { onRefresh() }
            }
            .onChange(of: menuState.exportTriggered) { _, triggered in
                if triggered { onExport() }
            }
            .onChange(of: menuState.printTriggered) { _, triggered in
                if triggered { onPrint() }
            }
            .onChange(of: menuState.themeToggleTriggered) { _, triggered in
                if triggered { onToggleTheme() }
            }
            .onChange(of: menuState.keyboardShortcutsTriggered) { _, triggered in
                if triggered { onShowKeyboardShortcuts() }
            }
            .onChange(of: menuState.navigateToTab) { _, tab in
                if let tab = tab { onGoToTab(tab) }
            }
    }
}

public extension View {
    /// Handle menu bar actions in this view
    func handleMenuBarActions(
        onSearch: @escaping () -> Void = {},
        onSettings: @escaping () -> Void = {},
        onFilters: @escaping () -> Void = {},
        onRefresh: @escaping () -> Void = {},
        onExport: @escaping () -> Void = {},
        onPrint: @escaping () -> Void = {},
        onToggleTheme: @escaping () -> Void = {},
        onGoToTab: @escaping (Int) -> Void = { _ in },
        onShowKeyboardShortcuts: @escaping () -> Void = {}
    ) -> some View {
        modifier(MenuBarActionHandler(
            onSearch: onSearch,
            onSettings: onSettings,
            onFilters: onFilters,
            onRefresh: onRefresh,
            onExport: onExport,
            onPrint: onPrint,
            onToggleTheme: onToggleTheme,
            onGoToTab: onGoToTab,
            onShowKeyboardShortcuts: onShowKeyboardShortcuts
        ))
    }
}

#endif

// MARK: - Cross-Platform Stubs

#if !os(macOS)
/// Stub for non-macOS platforms
public struct DesktopMenuService {
    public static var isDesktop: Bool { false }
}
#endif
